crates/kubelet/src/kubelet.rs

```diff
@@ -162,6 +141,107 @@ impl<P> Clone for Kubelet<P> {
     }
 }
 
+/// Listens for updates to pods on this node and forwards them to queue.
+async fn start_pod_informer<P: 'static + Provider + Sync + Send>(
+    client: kube::Client,
+    node_name: String,
+    mut queue: PodQueue<P>,
+) -> anyhow::Result<()> {
+    let node_selector = format!("spec.nodeName={}", node_name);
+    let params = ListParams {
+        field_selector: Some(node_selector),
+        ..Default::default()
+    };
+    let api = Api::<KubePod>::all(client);
+    let informer = Informer::new(api).params(params);
+    loop {
+        // TODO Should these results be ignored if Err? It seems reasonable to just keep trying,
+        // especially since this function is critical for a graceful shutdown.
+        let mut stream = informer.poll().await?.boxed();
+        while let Some(event) = stream.try_next().await? {
+            debug!("Handling Kubernetes pod event: {:?}", event);
+            match queue.enqueue(event).await {
+                Ok(()) => debug!("Enqueued event for processing"),
+                Err(e) => warn!("Error enqueuing pod event: {}", e),
+            };
+        }
+    }
+}
+
+/// Periodically renew node lease. Exits if signal is caught.
+async fn start_node_updater(
+    signal: Arc<AtomicBool>,
+    client: kube::Client,
+    node_name: String,
+) -> anyhow::Result<()> {
+    let sleep_interval = std::time::Duration::from_secs(10);
+    while !signal.load(Ordering::Relaxed) {
+        update_node(&client, &node_name).await;
+        tokio::time::delay_for(sleep_interval).await;
+    }
+    Ok(())
+}
+
+/// Checks for shutdown signal and cleans up resources gracefully.
+async fn start_signal_handler(
+    signal: Arc<AtomicBool>,
+    client: kube::Client,
+    node_name: String,
+) -> anyhow::Result<()> {
+    let duration = std::time::Duration::from_millis(100);
+    loop {
+        if signal.load(Ordering::Relaxed) {
+            warn!("Signal caught.");
+            drain_node(&client, &node_name).await?;
+            delete_node(&client, &node_name).await?;
```

Comment:
If I remember correctly, the normal kubelet doesn't deregister itself from the API. I think cordoning would be ok, but deleting probably isn't needed

Reply:
Hmmm, ok. I'm fine with this but I think then I would like to look at the re-registration functionality because I have been having issues with that.

Reply:
So right now the re-registration is a hack because we don't have TLS bootstrapping and can't update the node status directly. Another community member is almost done with that (we are waiting to cut 0.2 for it), so this should be greatly improved soon-ish

Reply:
Ok, sounds good.